src/lib/statistics/join_graph_statistics_cache.hpp

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <map>
+#include <unordered_map>
+
+#include "boost/dynamic_bitset.hpp"
+
+#include "expression/abstract_expression.hpp"
+
+namespace opossum {
+
+class TableStatistics;
+class JoinGraph;
+
+/**
+ * Cache of TableStatistics for LQPs consisting exclusively of JoinNodes, PredicateNodes and
+ * <Vertex Nodes>. That is, this is a cache for statistics associated with LQPs representing a subset of a JoinGraph.
+ *
+ * The key of the cache is a bitmask (with each bit representing a predicate or a vertex node).
+ *
+ * This cache exists primarily to aid the performance of the JoinOrderingRule.
+ * The JOR frequently request statistics for different plans consisting of the same set of Join and Scan predicates.
+ */
+class JoinGraphStatisticsCache {
+ public:
+  using Bitmask = boost::dynamic_bitset<>;
+  using VertexIndexMap = std::unordered_map<std::shared_ptr<AbstractLQPNode>, size_t>;
+  using PredicateIndexMap = ExpressionUnorderedMap<size_t>;
+
+  static JoinGraphStatisticsCache from_join_graph(const JoinGraph& join_graph);
+
+  JoinGraphStatisticsCache(VertexIndexMap&& vertex_indices, PredicateIndexMap&& predicate_indices);
+
+  /**
+   * Try to build a bitmask (aka cache key) from an LQP. This will either return the bitmask or std::nullopt, if
+   * Predicates or Vertices not registered in _vertex_indices and _predicate_indices are encountered.
+   */
+  std::optional<Bitmask> bitmask(const std::shared_ptr<AbstractLQPNode>& lqp) const;
+
+  /**
+   * Retrieve the cached statistics associated with @param bitmask. The order of columns in the returned TableStatistics
+   * will be as specified by @param requested_column_order. Returns nullptr if no cache entry exists for the specified
+   * bitmask.
```

Comment:
I found this way simpler than having the cache return a statistics object + a column order which the caller then has to reorganize. Tacked on or not, I'd leave it as it is. Caching statistics for JoinGraphs is fundamentally something that needs to deal with column reordering as different LQPs representing the same join graph will have different column orders.

Reply:
We discussed this in person and while there is also the possibility of `JoinGraphStatisticsCache::get()` returning a `CacheEntry`, which in turn might have a function `CacheEntry::get_statistics_with_column_order(....)` I like the current interface better since it makes cache set/get a single step respectively. I think in our discussion we concluded that this is acceptable.